### PR TITLE
refactor: Refactor ImplicitHttpApiPlugin and ImplicitRestApiPlugin

### PR DESCRIPTION
**Pull request by aahung**
### Description of changes

- combine duplicate code or similar code
- convert to constants if something is constant
- add types
- correct variable namings
- use generic classes

### Checklist

- [ ] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [ ] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/translator/output/error_api_swagger_integration_with_getatt_intrinsic_api_id.json [hunk 1]
```diff
@@ -1,3 +1,3 @@
 {
-  "errorMessage": "Invalid Serverless Application Specification document. Number of errors found: 1. Resource with id [HtmlFunction] is invalid. Event with id [GetHtml] is invalid. Api Event's RestApiId must be a string referencing an Api in the same template."
+  "errorMessage": "Invalid Serverless Application Specification document. Number of errors found: 1. Resource with id [HtmlFunction] is invalid. Event with id [GetHtml] is invalid. Property 'RestApiId' should be a string."
```
Comment (aahung) [inline tests/translator/output/error_api_swagger_integration_with_getatt_intrinsic_api_id.json:2]: The "in the same template" is not checked here, we have another test to cover it: https://github.com/aws/serverless-application-model/blob/develop/tests/translator/output/error_http_api_event_invalid_api.json

**Review by ssenchenko — APPROVED** [review]

File: samtranslator/validator/value_validator.py [hunk 2]
```diff
@@ -86,11 +86,11 @@ def to_be_a_list_of(self, expected_type: ExpectedType, message: Optional[str] =
...
 
-    def to_be_a_string(self, message: Optional[str] = "") -> T:
-        return self.to_be_a(ExpectedType.STRING, message)
+    def to_be_a_string(self, message: Optional[str] = "") -> str:
+        return cast(str, self.to_be_a(ExpectedType.STRING, message))
 
-    def to_be_an_integer(self, message: Optional[str] = "") -> T:
-        return self.to_be_a(ExpectedType.INTEGER, message)
+    def to_be_an_integer(self, message: Optional[str] = "") -> str:
+        return cast(str, self.to_be_a(ExpectedType.INTEGER, message))
```
Comment (acristin) [inline samtranslator/validator/value_validator.py:93]: I'm confused by the name and return type

Comment (aahung) [reply samtranslator/validator/value_validator.py:93]: this method raise exception if the value to validate is not an integer, and return the value "as an integer" to make mypy know it is now an integer.

(I immediately cannot read this sentence after I finished it lol)

Comment (aahung) [reply samtranslator/validator/value_validator.py]: @acristin added more docsrtring

**Review by hoffa — APPROVED** [review]

File: samtranslator/validator/value_validator.py (hunk 2 above)
Comment (acristin) [reply samtranslator/validator/value_validator.py]: >  and return the value "as an integer"

That's where my confusion comes from; the method returns `str`

Comment (aahung) [reply samtranslator/validator/value_validator.py]: yeah, it either return a `str` (if the value is a string, happy case), or raise excepion

Comment (aahung) [reply samtranslator/validator/value_validator.py]: fixed... I see where is the confusion

**Review by acristin — APPROVED** [review]
lgtm

**Pull request merged**